dolfin_adjoint_common/blocks/assembly.py

```diff
@@ -18,6 +18,41 @@ def __init__(self, form, ad_block_tag=None):
     def __str__(self):
         return f"assemble({ufl2unicode(self.form)})"
 
+    def compute_action_adjoint(self, adj_input, arity_form, form=None, c_rep=None, space=None, dform=None):
+        if arity_form == 0:
+            if dform is None:
+                dc = self.backend.TestFunction(space)
+                dform = self.backend.derivative(form, c_rep, dc)
+            dform_vector = self.compat.assemble_adjoint_value(dform)
+            return adj_input * dform_vector, dform
+        elif arity_form == 1:
+            if dform is None:
+                dc = self.backend.TrialFunction(space)
+                dform = self.backend.derivative(form, c_rep, dc)
+            # Get the Function
+            adj_input = adj_input.function
+            # Symbolic operators such as action/adjoint require derivatives to have been expanded beforehand.
+            # However, UFL doesn't support expanding coordinate derivatives of Coefficients in physical space,
+            # implying that we can't symbolically take the action/adjoint of the Jacobian for SpatialCoordinates.
+            # -> Workaround: Apply action/adjoint numerically (using PETSc).
+            if not isinstance(c_rep, self.backend.SpatialCoordinate):
+                # Symbolically compute: (dform/dc_rep)^* * adj_input
+                adj_output = self.backend.action(self.backend.adjoint(dform), adj_input)
+                adj_output = self.compat.assemble_adjoint_value(adj_output)
+            else:
+                # Get PETSc matrix
+                dform_mat = self.compat.assemble_adjoint_value(dform).petscmat
+                # Adjoint (hermitian transpose)
+                dform_mat.hermitianTranspose()
+                # Action
+                adj_output = self.backend.Function(space)
+                with adj_input.dat.vec_ro as v_vec:
+                    with adj_output.dat.vec as res_vec:
+                        dform_mat.mult(v_vec, res_vec)
+            return adj_output, dform
```

Comment:
It would be good to have confirmation of this workaround. In terms of the implementation, the SpatialCoordinate case (l. 44-51) should simplify with the upcoming dual space changes in UFL which will make that case as the normal case (cf. line 40-41) except that we will call `Adjoint/Action` instead of `adjoint/action`.

Unlike to the 0-form case, the adjoint of 1-forms requires doing symbolic manipulation with the Jacobian. UFL doesn't support expanding coordinate derivatives of `Coefficient`s in physical space which leads to assembling the 2-form instead of symbolically treating it to only assemble a 1-form at the end (cf. line 40-41). I am not fully happy with this but I don't see another solution at the moment.

Reply:
I think this will actually work. I agree it's a bit disgusting but let's not get distracted by this.

Can you put in a test that's got a space derivative?

Reply:
Space derivatives are systematically tested for any assembled forms as they always turn up in the dependency. I have updated the initial control value of `test_assemble_1_forms_adjoint` to make that derivative bigger and therefore more sensitive to testing.